juju/model.py

```diff
@@ -0,0 +1,210 @@
+            return False
+        return self.name == other.name
+
+
+class RelationMap(Mapping):
+    """Map of relation names to lists of Relation instances."""
+    def __init__(self, relation_names, local_app, local_unit, backend=None):
+        self._relation_names = relation_names
+        self._local_app = local_app
+        self._local_unit = local_unit
+        self._backend = backend or ModelBackend()
+
+    def __hash__(self):
+        return id(self)
+
+    def __eq__(self, other):
+        return self is other
+
+    def __iter__(self):
+        return iter(self._relation_names)
+
+    def __len__(self):
+        return len(self._relation_names)
+
+    @lru_cache(maxsize=None)
+    def __getitem__(self, relation_name):
+        # Lazy-load the list of relations for this relation name.
+        relation_ids = self._backend.relation_ids(relation_name)
+        return [Relation(relation_name, relation_id, self._local_app, self._local_unit, self._backend) for relation_id in relation_ids]
+
+
+class Relation:
+    def __init__(self, relation_name, relation_id, local_app, local_unit, backend=None):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._backend = backend or ModelBackend()
+        self.data = RelationData(self.relation_name, relation_id, local_app, local_unit, self._backend)
+
+    @property
+    @lru_cache(maxsize=None)
+    def apps(self):
+        # Lazy-load the apps on this relation.
+        unit_names = self._backend.relation_list(self.relation_id)
+        apps = {}
+        for unit_name in unit_names:
+            app_name = unit_name.split('/')[0]
+            app = apps.get(app_name)
+            if not app:
+                app = apps[app_name] = Application(app_name, [])
+            app.units.append(Unit(unit_name))
+        return list(apps.values())
+
+
+class RelationData(Mapping):
+    def __init__(self, relation_name, relation_id, local_app, local_unit, backend=None):
+        self.relation_name = relation_name
+        self.relation_id = relation_id
+        self._local_app = local_app
+        self._local_unit = local_unit
+        self._backend = backend or ModelBackend()
+
+    def __hash__(self):
+        return id(self)
+
+    def __eq__(self, other):
+        return self is other
+
+    @property
+    @lru_cache(maxsize=None)
+    def _members(self):
+        # Lazy-load the list of members for this relation.
+        members = set()
+        members.add(self._local_app)
+        members.add(self._local_unit)
+        unit_names = self._backend.relation_list(self.relation_id)
+        apps = {}
+        for unit_name in unit_names:
+            app_name = unit_name.split('/')[0]
+            app = apps.get(app_name)
+            if not app:
+                app = apps[app_name] = Application(app_name, [])
+                members.add(app)
+            unit = Unit(unit_name)
+            members.add(unit)
+            app.units.append(unit)
```

Comment:
This logic doesn't seem to belong here. I may be off since I didn't put much thinking on this, but this is relation.data.. we shouldn't have to load the list of members in a relation inside its data field.

Reply:
This is exclusively so that we can quickly detect in Python if an invalid key is accessed (e.g., a `Unit` from an unrelated relation) and raise an appropriate `KeyError`.  The other option is to always call `relation-get` and detect and possibly introspect errors to figure out which ones should be mapped to `KeyError` and which should be surfaced via some other error.

Maybe changing this from `_members()` to `keys()` would make it seem more reasonable?